modules/renter/repairscanner.go

```diff
@@ -0,0 +1,333 @@
+
+		// Fill out the set of unused hosts.
+		newUnfinishedChunks[i].pieceUsage = make([]bool, f.erasureCode.NumPieces())
+		newUnfinishedChunks[i].unusedHosts = make(map[string]struct{})
+		for host := range hosts {
+			newUnfinishedChunks[i].unusedHosts[host] = struct{}{}
+		}
+	}
+
+	// Iterate through the contracts of the file and mark which hosts are
+	// already in use for the chunk. As you delete hosts from the 'unusedHosts'
+	// map, also increment the 'piecesCompleted' value.
+	saveFile := false
+	for fcid, fileContract := range f.contracts {
+		// TODO: Need to figure out whether this contract line is still being
+		// used. And even worse, need to figure out whether this particular
+		// piece is still available in the contract, as hosts may drop
+		// particular pieces or lose particular drives, resulting in the
+		// contract line continuing to be valid but the data being gone.
+
+		// Convert the FileContractID into a host pubkey using the host pubkey
+		// lookup.
+		hpk, exists := fcidToHPK[fileContract.ID]
+		if !exists {
+			// File contract does not seem to be part of the host anymore.
+			// Delete this contract and mark the file to be saved.
+			delete(f.contracts, fcid)
+			saveFile = true
+			continue
+		}
+
+		// Mark the chunk set based on the pieces in this contract.
+		for _, piece := range fileContract.Pieces {
+			_, exists := newUnfinishedChunks[piece.Chunk].unusedHosts[hpk.String()]
+			nonRedundantPiece := newUnfinishedChunks[piece.Chunk].pieceUsage[piece.Piece]
+			if exists && nonRedundantPiece {
+				newUnfinishedChunks[piece.Chunk].pieceUsage[piece.Piece] = true
+				newUnfinishedChunks[piece.Chunk].piecesCompleted++
+				delete(newUnfinishedChunks[piece.Chunk].unusedHosts, hpk.String())
+			} else if exists {
+				// TODO / NOTE: This host has a piece, but it's the same piece
+				// that another host has. We may want to take action (such as
+				// deleting this piece from this host) because of this
+				// inefficiency.
+			}
+		}
+	}
+	// If 'saveFile' is marked, it means we deleted some dead contracts and
+	// cleaned up the file a bit. Save the file to clean up some space on disk
+	// and prevent the same work from being repeated after the next restart.
+	//
+	// TODO / NOTE: This process isn't going to make sense anymore once we
+	// switch to chunk-based saving.
+	if saveFile {
+		err := r.saveFile(f)
+		if err != nil {
+			r.log.Println("error while saving a file after pruning some contracts from it:", err)
+		}
+	}
+
+	// Iterate through the set of newUnfinishedChunks and remove any that are
+	// completed.
+	totalIncomplete := 0
+	for i := 0; i < len(newUnfinishedChunks); i++ {
+		if newUnfinishedChunks[i].piecesCompleted < newUnfinishedChunks[i].piecesNeeded {
+			newUnfinishedChunks[totalIncomplete] = newUnfinishedChunks[i]
+			newUnfinishedChunks[i].progress = float64(newUnfinishedChunks[i].piecesCompleted) / float64(newUnfinishedChunks[i].piecesNeeded)
+			totalIncomplete++
+		}
+	}
```

Comment:
IIUC, this could be:
```go
incompleteChunks := newUnfinishedChunks[:0]
for _, uc := range newUnfinishedChunks {
    if uc.pieceCompleted < uc.piecesNeeded {
        uc.progress = float64(uc.piecesCompleted) / float64(uc.piecesNeeded)
        incompleteChunks = append(incompleteChunks, uc)
    }
}
return incompleteChunks
```